### ISSUE TITLE
function called in while condition reported as coming from a different (for) loop inside the while body

### ISSUE THREAD

**Issue by demerphq** [Perl/perl5#20493]
**Description**
In https://github.com/Perl/perl5/pull/20492 you can see some dev/test code I was tinkering with to understand ParseXS.pm, in that code you can see a callstack reported by Carp::confess() which grossly misreports the line a given function is called on. The general structure of the code is
```
while ($ob->method()) { 
    ...

    for (keys %$thing) { 
        ...
    }
    ....
}
```

For some reason 'method' is being reported as coming from the for() loop. (See the draft PR for the exact details).     

**Steps to Reproduce**
See #20492 - I havent worked out a reduction yet. 

**Expected behavior**
I expect the method call to be reported as coming from the line it is on. I could sortof understand any line *associated* with the while() loop, but not a line from a totally different piece of code.

**Perl configuration**
Not bothering with a full -v, see #20492 for exact sha1 and etc.
```
./Configure -Dusethreads -Doptimize=-g -d -Dusedevel -Dcc=ccache\ gcc -Dld=gcc -DDEBUGGING
```

**Comment by iabyn** [comment]
On Mon, Nov 07, 2022 at 01:28:22AM -0800, Yves Orton wrote:
> **Description**
> In https://github.com/Perl/perl5/pull/20492 you can see some dev/test code I was tinkering with to understand ParseXS.pm, in that code you can see a callstack reported by Carp::confess() which grossly misreports the line a given function is called on. The general structure of the code is
> ```
> while ($ob->method()) { 
>     ...
> 
>     for (keys %$thing) { 
>         ...
>     }
>     ....
> }
> ```
> 
> For some reason 'method' is being reported as coming from the for() loop. (See the draft PR for the exact details).     

I don't know whether its the same as the issue you're seeing, but we
often see such artifacts with while loops, due to perl using the line
number from the last COP seen. For example in this loop, notice how the
first warning comes from the first line of the while loop, the rest from
the last line:

use warnings;
my $x = 4;
my $y;
while ($x > $y) {
    $x--;
}

Use of uninitialized value $y in numeric gt (>) at /home/davem/tmp/p line 5.
Use of uninitialized value $y in numeric gt (>) at /home/davem/tmp/p line 6.
Use of uninitialized value $y in numeric gt (>) at /home/davem/tmp/p line 6.
Use of uninitialized value $y in numeric gt (>) at /home/davem/tmp/p line 6.
Use of uninitialized value $y in numeric gt (>) at /home/davem/tmp/p line 6.


-- 
"There's something wrong with our bloody ships today, Chatfield."
    -- Admiral Beatty at the Battle of Jutland, 31st May 1916.

**Comment by demerphq** [comment]
Yeah, im aware of that oddness, but the thing about the case in this ticket is that its neither!
